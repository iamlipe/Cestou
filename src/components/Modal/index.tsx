import React, {useRef} from 'react';
import styled from 'styled-components/native';
import {ModalProps} from 'react-native';
import {SvgProps} from 'react-native-svg';
import {GestureHandlerRootView} from 'react-native-gesture-handler';
import {useNavigation} from '@react-navigation/native';

import Button from '@/components/Button';

interface Props extends ModalProps {
  title?: string;
  icon?: React.FC<SvgProps>;
  onPress?: () => void;
}

const Modal = ({title, icon: Icon, onPress, ...rest}: Props) => {
  const {goBack} = useNavigation();

  return (
    <GestureHandlerRootView>
      <StyledModal animationType="fade" transparent {...rest}>
        <StyledBackgroundModal>
          <StyledContent>
            {Icon && (
              <StyledCircle>
                <Icon width={60} height={60} />
              </StyledCircle>
            )}
            {title && <StyledTitle>{title}</StyledTitle>}

            {onPress && (
              <StyledContainerButtons>
                <StyledCancelButton
                  buttonColor="text_only"
                  textColor="primary"
                  size="medium"
                  title="Cancelar"
                  onPress={onPress}
                />

                <StyledConfirmButton
                  size="medium"
                  title="Confirmar"
                  onPress={() => {
                    onPress();
                    goBack();
                  }}
                />
              </StyledContainerButtons>
            )}
          </StyledContent>
        </StyledBackgroundModal>
      </StyledModal>
    </GestureHandlerRootView>
  );
};

const StyledModal = styled.Modal``;

const StyledBackgroundModal = styled.View`
  flex: 1;
  justify-content: center;
  align-items: center;
  background-color: ${({theme}) => theme.colors.BACKGROUND_MODAL};
`;

const StyledContent = styled.View`
  width: 80%;
  align-items: center;
  padding: 16px 32px;
  border-radius: 4px;
  background-color: ${({theme: {colors}}) => colors.BACKGROUND};
`;

const StyledTitle = styled.Text`
  margin: 16px 0 24px 0;
  text-align: center;
  line-height: 26px;
`;

const StyledContainerButtons = styled.View`
  width: 80%;
  flex-direction: row;
  justify-content: space-between;
`;

const StyledCircle = styled.View`
  width: 100px;
  height: 100px;
  align-items: center;
  justify-content: center;
  background-color: ${({theme}) => theme.colors.SECUNDARY_500};
  border-radius: 50px;
`;

const StyledCancelButton = styled(Button)``;

const StyledConfirmButton = styled(Button)``;

export default Modal;
