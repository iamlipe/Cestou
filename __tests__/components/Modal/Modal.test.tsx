import React from 'react';
import {
  renderWithThemeProvider,
  matchSnapshotWithProvider,
} from '@/config/functions/Testing';
import {SvgProps} from 'react-native-svg';

import IconMoldal from '@/assets/svgs/vegetable.svg';

import Modal from '@/components/Modal';

describe('Modal', () => {
  test('should rende correctly', () => {
    renderWithThemeProvider(<Modal />);
  });

  test('should matches snapshot', () => {
    matchSnapshotWithProvider(<Modal />);
  });

  test('should render title', () => {
    const modal = renderWithThemeProvider(<Modal title="some title" />);

    const title = modal.getByText(/some title/i);

    expect(title).toBeTruthy();
  });

  test('shoudl render Icon', () => {
    const modal = renderWithThemeProvider(
      <Modal icon={IconMoldal as React.FC<SvgProps>} />,
    );

    const icon = modal.getByTestId('icon-modal');

    expect(icon).toBeTruthy();
  });
});
