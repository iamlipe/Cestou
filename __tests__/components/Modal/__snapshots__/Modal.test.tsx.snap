// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should matches snapshot 1`] = `
<View>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    style={
      Array [
        Object {},
      ]
    }
    transparent={true}
    visible={true}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#00000026",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-end",
              "backgroundColor": "#FFFFFF",
              "borderRadius": 4,
              "paddingBottom": 16,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 16,
              "width": "80%",
            },
          ]
        }
      />
    </View>
  </Modal>
</View>
`;
